### ISSUE TITLE
'reindent all lines' and formatter have different formatting strategies

### ISSUE THREAD

**Issue by aeschli** [microsoft/vscode#19140]
Testing #19091
In a typescript file:
```
class B {
	public foo() {
		console.log('hello');
		switch (parseInt(this.toString())) {
			case 1: 
				break;
			case 2:
				break;
		}
	}
}
````
- use 'Reindent All Lines'
- Indenting of the 'break' statements is changed to align to the case statements
- Use the formatter: break statements are indented again

It would be better if formatter and the reindent command use the same indentation strategy.

**Comment by rebornix** [comment]
This is because that our indentation rules for TypeScript is not complete, the one we are using is 

```
decreaseIndentPattern: /^(.*\*\/)?\s*\}.*$/,
increaseIndentPattern: /^.*\{[^}"']*$/
```

The one TextMate or other editors are using is like

```
decreaseIndentPattern: /^\s*((?!\S.*\/[*]).*[*]\/\s*)?[})\]]|^\s*(case\b.*|default):\s*(\/\/.*|\/[*].*[*]\/\s*)?$/,
increaseIndentPattern: /(\{[^}"']*|\([^)"']*|\[[^\]"']*|^\s*(\{\}|\(\)|\[\]|(case\b.*|default):))\s*(\/\/.*|\/[*].*[*]\/\s*)?$/,
indentNextLinePattern: /^\s*(for|while|if|else)\b(?!.*[;{}]\s*(\/\/.*|\/[*].*[*]\/\s*)?$)/,
unIndentedLinePattern: /^(?!.*([;{}]|\S:)\s*(\/\/.*|\/[*].*[*]\/\s*)?$)(?!.*(\{[^}"']*|\([^)"']*|\[[^\]"']*|^\s*(\{\}|\(\)|\[\]|(case\b.*|default):))\s*(\/\/.*|\/[*].*[*]\/\s*)?$)(?!^\s*((?!\S.*\/[*]).*[*]\/\s*)?[})\]]|^\s*(case\b.*|default):\s*(\/\/.*|\/[*].*[*]\/\s*)?$)(?!^\s*(for|while|if|else)\b(?!.*[;{}]\s*(\/\/.*|\/[*].*[*]\/\s*)?$))/
```

But we still have good experience with TypeScript as we allow extensions to customize `onEnterRules`, the indentation adjusts correctly for most of the case when users press enter.

Right now `Reindent lines` only read indentation rules and doesn't take onEnter rules into consideration as at least for TypeScript there are a few overlap between the second indentation rules set and onEnterRules.

I suggest to make ReIndent Lines honor onEnterRules and remove unnecessary onEnterRules from TypeScript. cc @mjbvz . Since it's kind of breaking change I'd like to postpone it to following iteration and work together with Matt.

**Comment by Nimzozo** [comment]
A `case` label is not a statement and should not be indented, but aligned with the `switch`.

**Comment by wuhkuh** [comment]
Other issues have been closed in favor of this one, but this issue has gone stale. Could you give us an update?

**Comment by mjbvz** [comment]
Changing switch/case indentation for the JS/ts formatter is tracked by https://github.com/Microsoft/TypeScript/issues/18682

**Comment by dotDeeka** [comment]
Why was this issue closed and points to an issue about switch/case indentation specifically. This is an issue that effects more than just switch. Jsdoc comments are formatted differently by reindent lines vs format, as well as jsx.

I expect all indentation commands to indent the same

**Comment by mjbvz** [comment]
There is pretty much zero reason to use `Reindent all lines` in js/ts files; use `format document`. `Reindent all lines` has a lot of bugs and limitations and we do not plan on investing in it further given its complexity. Using `Format document` is the solution for js/ts files

**Comment by dotDeeka** [comment]
So, every couple of months I keep coming back to vscode hoping to be able to use it. There is a lot to like with vscode. However, the one thing that keeps getting to me is how indentation works. I stumbled back upon this issue I commented on a couple of months ago.

While I agree `Reindent all lines` is probably useless in favor of `Format document`, what I want to do is have `reindent current line` indent correctly.

Let me explain where I'm coming from on this. I am an avid emacs user, and in emacs the tab key will auto indent the current line. I changed my tab key in my vscode config to execute `reindent current line` expecting to achieve this behavior, but given the unexpected result, it doesn't work for me. I guess what I really wish is there was a `format current line` in addition to the already existing `format selection` and `format document`
